src/Misc/Util.h

```diff
@@ -0,0 +1,116 @@
+/*
+    Util.h - generic helpers and abbreviations
+
+    Copyright 2024,  Ichthyostega
+
+    This file is part of yoshimi, which is free software: you can redistribute
+    it and/or modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation; either version 2 of the License,
+    or (at your option) any later version.
+
+    yoshimi is distributed in the hope that it will be useful, but WITHOUT ANY
+    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+    FOR A PARTICULAR PURPOSE.  See the GNU General Public License (version 2
+    or later) for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with yoshimi.  If not, see <http://www.gnu.org/licenses/>.
+
+*/
+
+#ifndef UTIL_H
+#define UTIL_H
+
+#include <string>
+
+namespace std {// forward declarations to avoid pervasive includes
+
+  template<typename T>
+  class allocator;
+  template<typename K, typename CMP, class ALLO>
+  class set;
+
```

Comment:
this doesn't build against libc++:  

```
/usr/include/c++/v1/set:1486:43: error: reference to 'set' is ambiguous
 1486 | using set _LIBCPP_AVAILABILITY_PMR = std::set<_KeyT, _CompareT, polymorphic_allocator<_KeyT>>;
      |                                           ^
/usr/include/c++/v1/set:564:28: note: candidate found by name lookup is 'std::__1::set'
  564 | class _LIBCPP_TEMPLATE_VIS set {
      |                            ^
/builddir/yoshimi-2.3.3/src/Misc/Util.h:31:9: note: candidate found by name lookup is 'std::set'
   31 |   class set;
      |         ^
```

just include `<algorithm>` and `<set>` like normal...

Reply:
Thanks for reporting.
Which version of the library / compiler are you using?

Btw, mindlessly including some header in the whole code base can significantly increase built times, and thus turnaround. OTOH, STL headers are typically well modularised and not that problematic.

Do you really also need `<algorithm>` to make it work in your setup?

Reply:
> Which version of the library / compiler are you using?

it's llvm libc++ 18 (and the compiler is clang 18, but only the stl matters for that)

> Do you really also need <algorithm> to make it work in your setup?

it's the correct header for `std::find`. it might work without it but that would be relying on a (transitive) include to define it, which is prone to breaking in the future when the stl headers get shuffled around and less things are automatically included, as libc++ happens to like doing every other release (ironically, so less stuff is automatically included, and compile times would improve where those things are not really needed)

Reply:
As pointed out in the related [issue #108](https://github.com/Yoshimi/yoshimi/issues/208), I made this optimisation _prematurely_ (since in similar situations I had seen quite significant gains). But in this case an actual measurement (with `#include <set>` everywhere) revealed no tangible impact, presumably because we are including several more heavyweight STL headers in many compilation units downstream.

In the recent Bugfix release v.2.3.3.1 we have now removed this problematic forward declaration and just use the includes in the regular way. Sorry for the breakage

Reply:
no worries :) thanks for the changes and your work on yoshimi!